woocommerce/woocommerce
# Optimize maybe_assign_default_product_cat to drop caches only if changes were made

Dear developers,

method maybe_assign_default_product_cat in AssignDefaultCategory class assigns product to a default category if it has no categories assigned. After that it performs cache flush, transient deletion and updating of term counts. It can however happen, that query does nothing -> all affected products have at least one category remaining. In this case the functions following it should not run, preserving caches.

### Changes proposed in this Pull Request:

Check if $wpdb->query returns number of rows affected > 0 and only then run wp_cache_flush,  delete_transient and wp_update_term_count_now

### How to test the changes in this Pull Request:

1. create a product and assign it two categories
2. delete one of the categories and check if wp_cache_flush was run (it should not and product should not be in Uncategorized)
3. delete second category and check if wp_cache_flush was run (it should and product should be in Uncategorized)

💡 Manual testing of this change is somewhat tricky, and it's best assessed by a developer using XDebug or other tools. However, the following snippet (which could be added to a mu-plugin) can be used to test when the term counts transient is cleared (it writes to a log that can be found in **WooCommerce ‣ Status ‣ Logs**):

```php
add_action( 'delete_transient_wc_term_counts', function () {
	wc_get_logger()->debug( 'Term counts transient deleted!' );
} );
```

<div align="center"><img src="https://github.com/user-attachments/assets/549ea6b0-b9d3-458a-93f8-fc7e95717b0e" width="600"></div>

This may be useful because term counts are cleared when the cache is flushed, as referenced in the testing instructions. It's not perfect, though, because there are various reasons that the terms count transient may be deleted besides the code changed here.

### Changelog entry


<details>

<summary>Optimize maybe_assign_default_product_cat to drop caches only if changes were made</summary>

#### Significance




#### Type




#### Message 
Improve performance of maybe_assign_default_product_cat by only dropping cache and term recounting if changes were made in the database 
</details>
